z3rro/src/prover.rs

```diff
@@ -21,6 +50,65 @@ pub enum ProveResult {
     Unknown(ReasonUnknown),
 }
 
+/// Execute swine on the file located at file_path
+fn execute_swine(file_path: &Path) -> Result<(SatResult, String), ProverCommandError> {
+    let output = Command::new("swine").arg(file_path).output();
+
+    match output {
+        Ok(output) => {
+            let stdout = String::from_utf8_lossy(&output.stdout);
+            let mut lines_buffer: VecDeque<&str>= stdout.lines().collect();
+            let first_line = lines_buffer.pop_front().ok_or(ProverCommandError::ParseError)?;
+            if first_line.trim().to_lowercase() == "unsat" {
+                Ok((SatResult::Unsat, "".to_string()))
+            } else if first_line.trim().to_lowercase() == "sat" {
+                let _last_line = lines_buffer.pop_back().ok_or(ProverCommandError::ParseError)?;
+                if _last_line.contains("SHA") {
+                    let cex = lines_buffer.iter().join("");
+                    Ok((SatResult::Sat, cex))
+                } else {
+                    Err(ProverCommandError::ParseError)
+                }
+            } else {
+                lines_buffer.push_front(first_line);
+                Ok((SatResult::Unknown, lines_buffer.iter().join("\n")))
+            }
+        }
+        Err(e) => Err(ProverCommandError::ProcessError(e)),
+    }
+}
+
+/// In order to execute the program, it is necessary to remove lines that
+/// contain a forall quantifier or the declaration of the exponential function (exp).
+fn remove_lines_for_swine(input: &str) -> String {
+    let mut output = String::new();
+    let mut tmp_buffer: VecDeque<char> = VecDeque::new();
+    let mut input_buffer: VecDeque<char> = input.chars().collect();
+    let mut cnt = 0;
+
+    while let Some(c) = input_buffer.pop_front() {
```

Comment:
Maybe add a comment that you collect the full string for a (possibly multi-line) SMTLIB command by looking for the corresponding closed parenthesis, while counting nested parenthesis.